File: poc/rect_mesh.py
"""Proof-of-concept rectangular decision mesh (GEOMETRY_DESIGN.md).

Implements the geometry layer only, on the unit square with integer dyadic
coordinates: anisotropic dyadic bisection (x- and y-cuts), the two weld
species (statistical weld / hanging node), closed-form constraint
resolution, 2:1 edge balance with welded byproducts, composite release
splits, and bilinear evaluation. Statistics are stubbed with plain least
squares so the adaptive demo can run; no gates, no tau, no pipeline.

Vertex states (monotone over a run, no coarsening in the PoC):
  hanging -> weld -> free
"""

import numpy as np

MAXL = 12
S = 1 << MAXL  # integer span of the unit square


class Vertex:
    __slots__ = ("ix", "iy", "state", "height", "birth_parents", "depth", "birth_origin")

    def __init__(self, ix, iy, depth, birth_parents=None, birth_origin="root"):
        self.ix, self.iy = ix, iy
        self.state = "weld" if birth_parents else "free"
        self.height = 0.0
        self.birth_parents = birth_parents or []  # [(Vertex, 0.5), (Vertex, 0.5)]
        self.depth = depth
        self.birth_origin = birth_origin

    @property
    def x(self):
        return self.ix / S

    @property
    def y(self):
        return self.iy / S


class Cell:
    __slots__ = ("x0", "y0", "x1", "y1", "lx", "ly", "children", "axis")

    def __init__(self, x0, y0, x1, y1, lx, ly):
        self.x0, self.y0, self.x1, self.y1 = x0, y0, x1, y1
        self.lx, self.ly = lx, ly
        self.children = None
        self.axis = None

    @property
    def leaf(self):
        return self.children is None

    def corner_keys(self):
        return [(self.x0, self.y0), (self.x1, self.y0),
                (self.x0, self.y1), (self.x1, self.y1)]

    def contains(self, ix, iy):
        return self.x0 <= ix <= self.x1 and self.y0 <= iy <= self.y1


class RectMesh:
    def __init__(self, balance=True):
        self.balance = balance
        self.verts = {}
        for ix in (0, S):
            for iy in (0, S):
                self.verts[(ix, iy)] = Vertex(ix, iy, depth=0)
        self.root = Cell(0, 0, S, S, 0, 0)
        self.leaves = [self.root]
        # hanging map: vertex key -> (master_key_a, master_key_b)
        self.hanging = {}

    # ---------------------------------------------------------- topology

    def _get_or_make(self, ix, iy, parents, depth, origin):
        v = self.verts.get((ix, iy))
        if v is None:
            v = Vertex(ix, iy, depth, birth_parents=parents, birth_origin=origin)
            self.verts[(ix, iy)] = v
        return v

    def split(self, cell, axis, origin="selected"):
        """Dyadic bisection. New chord-endpoint vertices are born welded
        (species decided by the hanging recomputation afterwards)."""
        assert cell.leaf
        d = cell.lx + cell.ly + 1
        if axis == "x":
            xm = (cell.x0 + cell.x1) // 2
            pa = [(self.verts[(cell.x0, cell.y0)], 0.5), (self.verts[(cell.x1, cell.y0)], 0.5)]
            pb = [(self.verts[(cell.x0, cell.y1)], 0.5), (self.verts[(cell.x1, cell.y1)], 0.5)]
            self._get_or_make(xm, cell.y0, pa, d, origin)
            self._get_or_make(xm, cell.y1, pb, d, origin)
            a = Cell(cell.x0, cell.y0, xm, cell.y1, cell.lx + 1, cell.ly)
            b = Cell(xm, cell.y0, cell.x1, cell.y1, cell.lx + 1, cell.ly)
        else:
            ym = (cell.y0 + cell.y1) // 2
            pa = [(self.verts[(cell.x0, cell.y0)], 0.5), (self.verts[(cell.x0, cell.y1)], 0.5)]
            pb = [(self.verts[(cell.x1, cell.y0)], 0.5), (self.verts[(cell.x1, cell.y1)], 0.5)]
            self._get_or_make(cell.x0, ym, pa, d, origin)
            self._get_or_make(cell.x1, ym, pb, d, origin)
            a = Cell(cell.x0, cell.y0, cell.x1, ym, cell.lx, cell.ly + 1)
            b = Cell(cell.x0, ym, cell.x1, cell.y1, cell.lx, cell.ly + 1)
        cell.children, cell.axis = (a, b), axis
        self.leaves.remove(cell)
        self.leaves += [a, b]
        if self.balance:
            self._rebalance()
        self._recompute_hanging()
        return a, b

    def _rebalance(self):
        """2:1 edge balance: leaves sharing an edge segment may differ by at
        most one level in the axis transverse to that edge. Balance splits
        are welded byproducts (origin='balance')."""
        changed = True
        while changed:
            changed = False
            for c in list(self.leaves):
                for n in list(self.leaves):
                    if n is c:
                        continue
                    if c.x1 == n.x0 or c.x0 == n.x1:  # vertical shared edge
                        if not (c.y0 < n.y1 and n.y0 < c.y1):
                            continue
                        if n.ly < c.ly - 1:
                            self._plain_split(n, "y")
                            changed = True
                            break
                    if c.y1 == n.y0 or c.y0 == n.y1:  # horizontal shared edge
                        if not (c.x0 < n.x1 and n.x0 < c.x1):
                            continue
                        if n.lx < c.lx - 1:
                            self._plain_split(n, "x")
                            changed = True
                            break
                if changed:
                    break

    def _plain_split(self, cell, axis):
        """split() without recursion into rebalance (called from it)."""
        saved, self.balance = self.balance, False
        try:
            self.split(cell, axis, origin="balance")
        finally:
            self.balance = saved

    def _recompute_hanging(self):
        """A vertex strictly inside a leaf's edge is hanging, constrained to
        that edge's endpoints (masters). Overrides weld state; never
        overrides free (promotion requires regularity first)."""
        self.hanging = {}
        for leaf in self.leaves:
            edges = [((leaf.x0, leaf.y0), (leaf.x1, leaf.y0)),
                     ((leaf.x0, leaf.y1), (leaf.x1, leaf.y1)),
                     ((leaf.x0, leaf.y0), (leaf.x0, leaf.y1)),
                     ((leaf.x1, leaf.y0), (leaf.x1, leaf.y1))]
            for (ax, ay), (bx, by) in edges:
                for (ix, iy), v in self.verts.items():
                    if (ix, iy) == (ax, ay) or (ix, iy) == (bx, by):
                        continue
                    on = (ay == by == iy and ax < ix < bx) or \
                         (ax == bx == ix and ay < iy < by)
                    if on:
                        assert v.state != "free", "free vertex became hanging"
                        v.state = "hanging"
                        self.hanging[(ix, iy)] = ((ax, ay), (bx, by))
        for key, v in self.verts.items():
            if v.state == "hanging" and key not in self.hanging:
                v.state = "weld"  # released by a neighbor split

    # ------------------------------------------------------- constraints

    def masters(self, v):
        """Current constraint pair for a non-free vertex. Hanging weights are
        the linear interpolation parameter along the coarse edge — exactly
        1/2 at the midpoint (the only steady state under 2:1 balance), but
        general positions occur transiently mid-rebalance."""
        if v.state == "hanging":
            a, b = self.hanging[(v.ix, v.iy)]
            (ax, ay), (bx, by) = a, b
            t = ((v.ix - ax) / (bx - ax)) if ax != bx else ((v.iy - ay) / (by - ay))
            return [(self.verts[a], 1.0 - t), (self.verts[b], t)]
        return v.birth_parents

    def closure(self, v):
        """Fully resolved master list {(free vertex, weight)} (GEOMETRY_DESIGN §3)."""
        if v.state == "free":
            return {(v.ix, v.iy): 1.0}
        out = {}
        for m, w in self.masters(v):
            for k, w2 in self.closure(m).items():
                out[k] = out.get(k, 0.0) + w * w2
        return out

    def resolve_heights(self):
        for v in sorted(self.verts.values(), key=lambda v: v.depth):
            if v.state != "free":
                v.height = sum(w * m.height for m, w in self.masters(v))

    def resolve_heights_hier(self, delta):
        """Hierarchical (surplus) parameterization: every vertex sits at its
        parent interpolation plus its own surplus; welds and hanging vertices
        have surplus structurally zero (delta only holds free keys)."""
        for v in sorted(self.verts.values(), key=lambda v: v.depth):
            base = sum(w * m.height for m, w in self.masters(v)) \
                if (v.birth_parents or v.state == "hanging") else 0.0
            v.height = base + delta.get((v.ix, v.iy), 0.0)

    def design_hier(self, xs, ys):
        """Design columns in the hierarchical basis: column of delta_k is the
        surface response to a unit surplus at k, transported through all
        descendants. Promotion adds a column here without moving any other
        (GEOMETRY_DESIGN §2b)."""
        keys = self.free_keys()
        saved = {k: self.verts[k].height for k in self.verts}
        cols = []
        for k in keys:
            self.resolve_heights_hier({k: 1.0})
            cols.append([self.eval(x, y) for x, y in zip(xs, ys)])
        for kk, h in saved.items():
            self.verts[kk].height = h
        return keys, np.array(cols).T

    # ------------------------------------------------------- promotion

    def promote(self, key):
        """Composite move: release splits until regular, then free (§2, §7.3)."""
        n_release = 0
        while self.verts[key].state == "hanging":
            (ax, ay), (bx, by) = self.hanging[key]
            leaf = next(l for l in self.leaves
                        if l.contains(*key) and l.contains(ax, ay) and l.contains(bx, by))
            self.split(leaf, "y" if ay != by else "x", origin="release")
            n_release += 1
        self.verts[key].state = "free"
        return n_release

    # ------------------------------------------------------- evaluation

    def find_leaf(self, x, y):
        ix = min(max(x, 0.0), 1.0) * S
        iy = min(max(y, 0.0), 1.0) * S
        c = self.root
        while not c.leaf:
            a, b = c.children
            c = a if (ix <= a.x1 if c.axis == "x" else iy <= a.y1) else b
        return c

    def eval(self, x, y):
        c = self.find_leaf(x, y)
        u = (x * S - c.x0) / (c.x1 - c.x0)
        w = (y * S - c.y0) / (c.y1 - c.y0)
        h00 = self.verts[(c.x0, c.y0)].height
        h10 = self.verts[(c.x1, c.y0)].height
        h01 = self.verts[(c.x0, c.y1)].height
        h11 = self.verts[(c.x1, c.y1)].height
        return (h00 * (1 - u) * (1 - w) + h10 * u * (1 - w)
                + h01 * (1 - u) * w + h11 * u * w)

    def eval_grid(self, n=96):
        g = (np.arange(n) + 0.5) / n
        self.resolve_heights()
        return np.array([[self.eval(x, y) for x in g] for y in g])

    # ------------------------------------------------------- least squares

    def free_keys(self):
        return [k for k, v in self.verts.items() if v.state == "free"]

    def design(self, xs, ys):
        """Design column per free coefficient via unit-height probes; exact
        through the constraint closure by construction."""
        keys = self.free_keys()
        saved = {k: self.verts[k].height for k in self.verts}
        cols = []
        for k in keys:
            for kk in self.verts:
                self.verts[kk].height = 0.0
            self.verts[k].height = 1.0
            self.resolve_heights()
            cols.append([self.eval(x, y) for x, y in zip(xs, ys)])
        for kk, h in saved.items():
            self.verts[kk].height = h
        return keys, np.array(cols).T

    def fit(self, xs, ys, zs, ridge=1e-8):
        keys, X = self.design(xs, ys)
        A = X.T @ X + ridge * np.eye(len(keys))
        b = X.T @ np.asarray(zs)
        h = np.linalg.solve(A, b)
        for k, hv in zip(keys, h):
            self.verts[k].height = hv
        self.resolve_heights()
        return h

File: core/rect_mesh.cpp
#include "rect_mesh.h"

#include <algorithm>
#include <cassert>
#include <cstdlib>
#include <stdexcept>

namespace rectmesh {

RectMesh::RectMesh() {
    for (int ix : {0, kS})
        for (int iy : {0, kS}) {
            auto v = std::make_unique<Vertex>();
            v->ix = ix;
            v->iy = iy;
            v->state = VState::Free;
            v->depth = 0;
            v->origin = Origin::Root;
            verts_[key_of(ix, iy)] = v.get();
            vert_store_.push_back(std::move(v));
        }
    auto c = std::make_unique<Cell>();
    c->x0 = 0; c->y0 = 0; c->x1 = kS; c->y1 = kS; c->lx = 0; c->ly = 0;
    root_ = c.get();
    cell_store_.push_back(std::move(c));
    leaves_.push_back(root_);
}

Vertex* RectMesh::get_or_make(int ix, int iy,
                              std::vector<std::pair<Vertex*, double>> parents,
                              int depth, Origin origin) {
    auto it = verts_.find(key_of(ix, iy));
    if (it != verts_.end()) return it->second;
    auto v = std::make_unique<Vertex>();
    v->ix = ix;
    v->iy = iy;
    v->state = VState::Weld;
    v->birth_parents = std::move(parents);
    v->depth = depth;
    v->origin = origin;
    Vertex* raw = v.get();
    verts_[key_of(ix, iy)] = raw;
    vert_store_.push_back(std::move(v));
    return raw;
}

void RectMesh::plain_split(Cell* cell, char axis, Origin origin) {
    assert(cell->leaf());
    const int d = cell->lx + cell->ly + 1;
    auto va = [&](int ix, int iy) { return verts_.at(key_of(ix, iy)); };
    Cell *a, *b;
    auto ca = std::make_unique<Cell>();
    auto cb = std::make_unique<Cell>();
    a = ca.get();
    b = cb.get();
    if (axis == 'x') {
        const int xm = (cell->x0 + cell->x1) / 2;
        get_or_make(xm, cell->y0,
                    {{va(cell->x0, cell->y0), 0.5}, {va(cell->x1, cell->y0), 0.5}},
                    d, origin);
        get_or_make(xm, cell->y1,
                    {{va(cell->x0, cell->y1), 0.5}, {va(cell->x1, cell->y1), 0.5}},
                    d, origin);
        *a = {cell->x0, cell->y0, xm, cell->y1, cell->lx + 1, cell->ly};
        *b = {xm, cell->y0, cell->x1, cell->y1, cell->lx + 1, cell->ly};
    } else {
        const int ym = (cell->y0 + cell->y1) / 2;
        get_or_make(cell->x0, ym,
                    {{va(cell->x0, cell->y0), 0.5}, {va(cell->x0, cell->y1), 0.5}},
                    d, origin);
        get_or_make(cell->x1, ym,
                    {{va(cell->x1, cell->y0), 0.5}, {va(cell->x1, cell->y1), 0.5}},
                    d, origin);
        *a = {cell->x0, cell->y0, cell->x1, ym, cell->lx, cell->ly + 1};
        *b = {cell->x0, ym, cell->x1, cell->y1, cell->lx, cell->ly + 1};
    }
    cell->child_a = a;
    cell->child_b = b;
    cell->axis = axis;
    cell_store_.push_back(std::move(ca));
    cell_store_.push_back(std::move(cb));
    for (size_t i = 0; i < leaves_.size(); ++i)
        if (leaves_[i] == cell) {
            leaves_[i] = leaves_.back();
            leaves_.pop_back();
            break;
        }
    leaves_.push_back(a);
    leaves_.push_back(b);
    ++topo_version_;
}

void RectMesh::split(Cell* cell, char axis, Origin origin) {
    plain_split(cell, axis, origin);
    if (balance_enabled_) rebalance();
    recompute_hanging();
}

void RectMesh::rebalance() {
    // 2:1 edge balance: leaves sharing an edge segment differ by at most
    // one level in the axis transverse to that edge. Balance splits are
    // welded byproducts.
    bool changed = true;
    while (changed) {
        changed = false;
        for (size_t i = 0; i < leaves_.size() && !changed; ++i) {
            Cell* c = leaves_[i];
            for (size_t j = 0; j < leaves_.size() && !changed; ++j) {
                Cell* n = leaves_[j];
                if (n == c) continue;
                const bool vshare = (c->x1 == n->x0 || c->x0 == n->x1) &&
                                    (c->y0 < n->y1 && n->y0 < c->y1);
                if (vshare && n->ly < c->ly - 1) {
                    plain_split(n, 'y', Origin::Balance);
                    changed = true;
                    break;
                }
                const bool hshare = (c->y1 == n->y0 || c->y0 == n->y1) &&
                                    (c->x0 < n->x1 && n->x0 < c->x1);
                if (hshare && n->lx < c->lx - 1) {
                    plain_split(n, 'x', Origin::Balance);
                    changed = true;
                    break;
                }
            }
        }
    }
}

void RectMesh::recompute_hanging() {
    hanging_.clear();
    for (Cell* leaf : leaves_) {
        const int ex[4][4] = {
            {leaf->x0, leaf->y0, leaf->x1, leaf->y0},
            {leaf->x0, leaf->y1, leaf->x1, leaf->y1},
            {leaf->x0, leaf->y0, leaf->x0, leaf->y1},
            {leaf->x1, leaf->y0, leaf->x1, leaf->y1}};
        for (const auto& e : ex) {
            const int ax = e[0], ay = e[1], bx = e[2], by = e[3];
            for (auto& kv : verts_) {
                Vertex* v = kv.second;
                if ((v->ix == ax && v->iy == ay) ||
                    (v->ix == bx && v->iy == by))
                    continue;
                const bool on =
                    (ay == by && v->iy == ay && ax < v->ix && v->ix < bx) ||
                    (ax == bx && v->ix == ax && ay < v->iy && v->iy < by);
                if (on) {
                    assert(v->state != VState::Free &&
                           "free vertex became hanging");
                    v->state = VState::Hanging;
                    hanging_[kv.first] = {key_of(ax, ay), key_of(bx, by)};
                }
            }
        }
    }
    for (auto& kv : verts_) {
        Vertex* v = kv.second;
        if (v->state == VState::Hanging && !hanging_.count(kv.first))
            v->state = VState::Weld;  // released by a neighbor split
    }
    ++topo_version_;
}

std::vector<std::pair<Vertex*, double>> RectMesh::masters(
    const Vertex* v) const {
    if (v->state == VState::Hanging) {
        const auto& ab = hanging_.at(key_of(v->ix, v->iy));
        Vertex* a = verts_.at(ab.first);
        Vertex* b = verts_.at(ab.second);
        double t;
        if (a->ix != b->ix)
            t = static_cast<double>(v->ix - a->ix) / (b->ix - a->ix);
        else
            t = static_cast<double>(v->iy - a->iy) / (b->iy - a->iy);
        return {{a, 1.0 - t}, {b, t}};
    }
    return v->birth_parents;
}

int RectMesh::promote(VKey key) {
    int n_release = 0;
    while (verts_.at(key)->state == VState::Hanging) {
        const auto& ab = hanging_.at(key);
        const int ax = key_ix(ab.first), ay = key_iy(ab.first);
        const int bx = key_ix(ab.second), by = key_iy(ab.second);
        Cell* target = nullptr;
        for (Cell* l : leaves_)
            if (l->contains(key_ix(key), key_iy(key)) && l->contains(ax, ay) &&
                l->contains(bx, by)) {
                target = l;
                break;
            }
        if (!target) throw std::runtime_error("promote: no constraining leaf");
        split(target, ay != by ? 'y' : 'x', Origin::Release);
        ++n_release;
    }
    verts_.at(key)->state = VState::Free;
    return n_release;
}

std::vector<VKey> RectMesh::free_keys() const {
    std::vector<VKey> out;
    for (auto& kv : verts_)
        if (kv.second->state == VState::Free) out.push_back(kv.first);
    return out;
}

void RectMesh::resolve_heights_hier(
    const std::unordered_map<VKey, double>& delta) {
    // depth order: masters always have strictly lower depth
    std::vector<Vertex*> order;
    order.reserve(verts_.size());
    for (auto& kv : verts_) order.push_back(kv.second);
    std::sort(order.begin(), order.end(),
              [](const Vertex* a, const Vertex* b) { return a->depth < b->depth; });
    for (Vertex* v : order) {
        double base = 0.0;
        if (!v->birth_parents.empty() || v->state == VState::Hanging)
            for (auto& mw : masters(v)) base += mw.second * mw.first->height;
        auto it = delta.find(key_of(v->ix, v->iy));
        v->height = base + (it == delta.end() ? 0.0 : it->second);
    }
}

const Cell* RectMesh::find_leaf(double x, double y) const {
    double ix = x * kS, iy = y * kS;
    if (ix < 0) ix = 0;
    if (ix > kS) ix = kS;
    if (iy < 0) iy = 0;
    if (iy > kS) iy = kS;
    const Cell* c = root_;
    while (!c->leaf())
        c = ((c->axis == 'x') ? (ix <= c->child_a->x1) : (iy <= c->child_a->y1))
                ? c->child_a
                : c->child_b;
    return c;
}

double RectMesh::eval(double x, double y) const {
    const Cell* c = find_leaf(x, y);
    const double u = (x * kS - c->x0) / (c->x1 - c->x0);
    const double w = (y * kS - c->y0) / (c->y1 - c->y0);
    const double h00 = verts_.at(key_of(c->x0, c->y0))->height;
    const double h10 = verts_.at(key_of(c->x1, c->y0))->height;
    const double h01 = verts_.at(key_of(c->x0, c->y1))->height;
    const double h11 = verts_.at(key_of(c->x1, c->y1))->height;
    return h00 * (1 - u) * (1 - w) + h10 * u * (1 - w) + h01 * (1 - u) * w +
           h11 * u * w;
}

void RectMesh::closure_of(Vertex* v, std::unordered_map<VKey, double>* acc,
                          double w) {
    if (v->state == VState::Free) (*acc)[key_of(v->ix, v->iy)] += w;
    if (!v->birth_parents.empty() || v->state == VState::Hanging)
        for (auto& mw : masters(v)) closure_of(mw.first, acc, w * mw.second);
}

const std::vector<std::pair<VKey, double>>& RectMesh::closure(Vertex* v) {
    if (closure_version_ != topo_version_) {
        closure_memo_.clear();
        closure_version_ = topo_version_;
    }
    auto it = closure_memo_.find(v);
    if (it != closure_memo_.end()) return it->second;
    std::unordered_map<VKey, double> acc;
    closure_of(v, &acc, 1.0);
    std::vector<std::pair<VKey, double>> out(acc.begin(), acc.end());
    return closure_memo_[v] = std::move(out);
}

void RectMesh::design_row(double x, double y,
                          const std::unordered_map<VKey, int>& key_index,
                          std::vector<std::pair<int, double>>* out) {
    out->clear();
    const Cell* c = find_leaf(x, y);
    const double u = (x * kS - c->x0) / (c->x1 - c->x0);
    const double w = (y * kS - c->y0) / (c->y1 - c->y0);
    const std::pair<VKey, double> corners[4] = {
        {key_of(c->x0, c->y0), (1 - u) * (1 - w)},
        {key_of(c->x1, c->y0), u * (1 - w)},
        {key_of(c->x0, c->y1), (1 - u) * w},
        {key_of(c->x1, c->y1), u * w}};
    std::unordered_map<int, double> acc;
    for (const auto& cw : corners) {
        if (cw.second == 0.0) continue;
        for (const auto& fw : closure(verts_.at(cw.first)))
            acc[key_index.at(fw.first)] += cw.second * fw.second;
    }
    out->assign(acc.begin(), acc.end());
}

void RectMesh::probe_column(VKey v0, const std::vector<double>& xs,
                            const std::vector<double>& ys,
                            std::vector<double>* out) {
    // influence I(v; v0) = [v==v0] + sum_m w * I(m; v0), memoized locally
    std::unordered_map<const Vertex*, double> infl;
    // depth order guarantees masters computed first
    std::vector<Vertex*> order;
    order.reserve(verts_.size());
    for (auto& kv : verts_) order.push_back(kv.second);
    std::sort(order.begin(), order.end(),
              [](const Vertex* a, const Vertex* b) { return a->depth < b->depth; });
    for (Vertex* v : order) {
        double s = (key_of(v->ix, v->iy) == v0) ? 1.0 : 0.0;
        if (!v->birth_parents.empty() || v->state == VState::Hanging)
            for (auto& mw : masters(v)) s += mw.second * infl[mw.first];
        infl[v] = s;
    }
    out->resize(xs.size());
    for (size_t i = 0; i < xs.size(); ++i) {
        const Cell* c = find_leaf(xs[i], ys[i]);
        const double u = (xs[i] * kS - c->x0) / (c->x1 - c->x0);
        const double w = (ys[i] * kS - c->y0) / (c->y1 - c->y0);
        (*out)[i] = infl[verts_.at(key_of(c->x0, c->y0))] * (1 - u) * (1 - w) +
                    infl[verts_.at(key_of(c->x1, c->y0))] * u * (1 - w) +
                    infl[verts_.at(key_of(c->x0, c->y1))] * (1 - u) * w +
                    infl[verts_.at(key_of(c->x1, c->y1))] * u * w;
    }
}

}  // namespace rectmesh
